core/src/main/java/org/apache/brooklyn/util/core/crypto/SecureKeys.java

```diff
@@ -78,11 +78,11 @@ public static X509Principal getX509PrincipalWithCommonName(String commonName) {
         return new X509Principal("" + "C=None," + "L=None," + "O=None," + "OU=None," + "CN=" + commonName);
     }
 
-    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible
+    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible. Closes the stream.
      * @throws IllegalStateException on errors, in particular {@link PassphraseProblem} if that is the problem */
     public static KeyPair readPem(InputStream input, final String passphrase) {
         // TODO cache is only for fallback "reader" strategy (2015-01); delete when Parser confirmed working
-        byte[] cache = Streams.readFully(input);
+        byte[] cache = Streams.readFullyAndClose(input);
```

Comment:
@neykov How about overloading `readPem` to take a byte array? It looks like that is the way it is used anyway.

And for the tests that call it, I wonder if we should add `byte[] ResourceUtils.getResourceAsBytes(String url)`. Again it feels like many callers of the existing `getResourceFromUrl` either want it as a string or as bytes.